### PR TITLE
Added in a RestrictedFunctionSpace class

### PR DESCRIPTION
**Pull request by emmarothwell1**
Add in a RestrictedFunctionSpace class.

This marks boundary nodes in the definition of the FunctionSpace with the goal of removing them from calculations such as in assembly. 
This currently only works for FunctionSpace objects, restricted mixed spaces and restrictions based on already defined RestrictedFunctionSpaces are still being worked on elsewhere. 
There are tests in tests/regression/test_restricted_function_space.py to check assembly and solve works. 
There is also a linked PyOP2 pull request at: [https://github.com/OP2/PyOP2/pull/716](url)

### CODE REVIEW

File: firedrake/cython/dmcommon.pyx [hunk 1]
```diff
@@ -2331,8 +2427,13 @@ def plex_renumbering(PETSc.DM plex,
                         CHKERR(DMLabelHasPoint(labels[l], p, &has_point))
                         if has_point:
                             PetscBTSet(seen, p)
-                            perm[lidx[l]] = p
-                            lidx[l] += 1
+                            if boundary_set and PetscBTLookup(seen_boundary, p) and l <= 1:
```
Comment (dham) [inline firedrake/cython/dmcommon.pyx:2430]: please don't use the variable name `l`. It's super confusing.

**Review by dham — CHANGES_REQUESTED** [review]
This is very close. Just a few minor issues.

**Comment by ksagiyam** [comment]
This PR https://github.com/firedrakeproject/firedrake/pull/3288 has just been merged. Rebasing will cause conflicts in `assemble.py`, so I suggest that you focus on fixing the boundary condition issue first. Once you resolve the BC issue, I can do rebasing.

**Comment by ksagiyam** [comment]
It seems the big change in assemble.py has successfully been merged into this branch. The PR on halo_exchange_sf was merged yesterday, so you can drop the commit you cherry picked, and merge `master`.

**Comment by ksagiyam** [comment]
Superseded https://github.com/firedrakeproject/firedrake/pull/3215.

**Pull request closed without merging**